include/ginkgo/core/matrix/fft.hpp

```diff
@@ -0,0 +1,308 @@
+ *
+ * For a power-of-two sizes $n_1, n_2, n_3$ with corresponding root of unity
+ * $\omega = e^{-2\pi i / (n_1 n_2 n_3)}$ for forward DFT and
+ * $\omega = e^{2 \pi i / (n_1 n_2 n_3)}$ for inverse DFT it computes
+ *
+ *     $$ x_{k_1 n_2 n_3 + k_2 n_3 + k_3} = \sum_{i_1=0}^{n_1-1}
+ *                     \sum_{i_2=0}^{n_2-1} \sum_{i_3=0}^{n_3-1}
+ *                     \omega^{i_1 k_1 + i_2 k_2 + i_3 k_3}
+ *                     b_{i_1 n_2 n_3 + i_2 n_3 + i_3} $$
+ *
+ * without normalization factors.
+ *
+ * @ingroup identity
+ * @ingroup mat_formats
+ * @ingroup LinOp
+ */
+class Fft3 : public EnableLinOp<Fft3>,
+             public EnableCreateMethod<Fft3>,
+             public Transposable {
+    friend class EnablePolymorphicObject<Fft3, LinOp>;
+    friend class EnableCreateMethod<Fft3>;
+
+public:
+    using EnableLinOp<Fft3>::convert_to;
+    using EnableLinOp<Fft3>::move_to;
+
+    using value_type = std::complex<double>;
+    using transposed_type = Fft3;
+
+    std::unique_ptr<LinOp> transpose() const override;
+
+    std::unique_ptr<LinOp> conj_transpose() const override;
+
+    bool is_inverse() const;
+
+protected:
+    /**
+     * Creates an empty Fourier matrix.
+     *
+     * @param exec  Executor associated to the matrix
+     */
+    explicit Fft3(std::shared_ptr<const Executor> exec)
+        : EnableLinOp<Fft3>(exec),
+          buffer_{exec},
+          size1_{},
+          size2_{},
+          size3_{},
+          inverse_{}
+    {}
+
+    /**
+     * Creates an Fourier matrix with the given dimensions.
+     *
+     * @param size  size of all FFT dimensions (must be a power of two)
+     */
+    Fft3(std::shared_ptr<const Executor> exec, size_type size)
+        : Fft3{exec, size, size, size}
+    {}
+
+    /**
+     * Creates an Fourier matrix with the given dimensions.
+     *
+     * @param size1  size of the first FFT dimension (must be a power of two)
+     * @param size2  size of the second FFT dimension (must be a power of two)
+     * @param size3  size of the third FFT dimension (must be a power of two)
+     * @param inverse  true to compute an inverse DFT instead of a normal DFT
+     */
+    Fft3(std::shared_ptr<const Executor> exec, size_type size1, size_type size2,
+         size_type size3, bool inverse = false)
+        : EnableLinOp<Fft3>(exec, dim<2>{size1 * size2 * size3}),
+          buffer_{exec},
+          size1_{size1},
+          size2_{size2},
+          size3_{size3},
+          inverse_{inverse}
+    {
+        GKO_ASSERT_IS_POWER_OF_TWO(size1);
+        GKO_ASSERT_IS_POWER_OF_TWO(size2);
+        GKO_ASSERT_IS_POWER_OF_TWO(size3);
+    }
+
+    void apply_impl(const LinOp *b, LinOp *x) const override;
+
+    void apply_impl(const LinOp *alpha, const LinOp *b, const LinOp *beta,
+                    LinOp *x) const override;
+
+private:
+    mutable Array<char> buffer_;
+    size_type size1_;
+    size_type size2_;
+    size_type size3_;
```

Comment:
I would prefer if this was a `std::array<size_type, 3>` (same for 2D).

Reply:
![neutral](https://user-images.githubusercontent.com/1693511/111533596-3ecdfb80-8767-11eb-8fb8-e5d6e9bffa98.jpg)